v2/yarpcthrift/codec.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpcthrift
+
+import (
+	"bytes"
+
+	"go.uber.org/yarpc/v2/yarpcerror"
+
+	"go.uber.org/thriftrw/protocol"
+	"go.uber.org/thriftrw/wire"
+	yarpc "go.uber.org/yarpc/v2"
+)
+
+type thriftCodec struct {
+	protocol   protocol.Protocol
+	enveloping bool
+
+	responder protocol.Responder
+}
+
+func newCodec(protocol protocol.Protocol, enveloping bool) *thriftCodec {
+	return &thriftCodec{
+		protocol:   protocol,
+		enveloping: enveloping,
+	}
+}
+
+// setResponder checks whether a responder has been set and sets it to one. An
+// instance of a thrift codec should not set a responder more than once; if
+// it does, it may not encode a response properly. We use this instead of
+// directly setting a responder to ensure that we only set it once.
+func (c *thriftCodec) setResponder(responder protocol.Responder) error {
+	if c.responder != nil {
+		return yarpcerror.InternalErrorf("tried to overwrite a responder for thrift codec")
+	}
+	c.responder = responder
+	return nil
+}
+
+func (c *thriftCodec) Decode(req *yarpc.Buffer) (interface{}, error) {
+	// TODO(mhp): This will be unnecessary when `yarpc.Buffer` implements `io.ReaderAt`
+	reader := bytes.NewReader(req.Bytes())
+
+	// Discover or choose the appropriate envelope
```

Comment:
What's the significance of prioritizing the `EnvelopeAgnosticProtocol` over the `c.enveloping` value? In other words, is it ever possible for both of these conditions to be true? 

I'm sure that this makes sense in general, but we might want to document a simple reason for that here.

Reply:
EnvelopeAgnosticProtocol is first because it should *always* be used. The only case where you need a hint is if you’re stuck with a prior version of thriftrw that does not implement `protocol.EnvelopeAgnosticProtocol`. If we were in a monorepo, we would delete everything but this first case.